src/node/txreconciliation.cpp

```diff
@@ -134,14 +202,119 @@ class TxReconciliationTracker::Impl
+    }
+
+    bool IsFanoutTarget(CSipHasher&& deterministic_randomizer,
+                        bool we_initiate, double limit,
+                        NodeId peer_id, const Wtxid& wtxid) EXCLUSIVE_LOCKS_REQUIRED(m_txreconciliation_mutex)
+    {
+        auto fanout_candidates = tx_fanout_targets_cache_data.find(wtxid);
+        if (fanout_candidates != tx_fanout_targets_cache_data.end()) {
+            return fanout_candidates->second.find(peer_id) != fanout_candidates->second.end();
+        }
+
+        // We use the pre-determined randomness to give a consistent result per transaction,
+        // thus making sure that no transaction gets "unlucky" if every per-peer roll fails.
+        deterministic_randomizer.Write(wtxid.ToUint256());
+
+        // The algorithm works as follows. We iterate through the peers (of a given direction)
+        // hashing them with the given wtxid, and sort them by this hash.
+        // We then consider top `limit` peers to be low-fanout flood targets.
+        // The randomness should be seeded with wtxid to return consistent results for every call.
+
+        // The integral part of `limit` is accounted in the higher 32 bits of the second element.
+        // The fractional part of `limit` is stored in the lower 32 bits, and then we check
+        // whether adding a random lower 32-bit value (first element) would end up modifying
+        // the higher bits.
+        const size_t targets_size = ((deterministic_randomizer.Finalize() & 0xFFFFFFFF) + uint64_t(limit * 0x100000000)) >> 32;
+
+        std::vector<std::pair<uint64_t, NodeId>> best_peers;
+        best_peers.reserve(m_states.size());
+
+        for (const auto& indexed_state : m_states) {
+            const auto cur_state = std::get_if<TxReconciliationState>(&indexed_state.second);
+            if (cur_state && cur_state->m_we_initiate == we_initiate) {
+                uint64_t hash_key = CSipHasher(deterministic_randomizer).Write(cur_state->m_k0).Finalize();
+                best_peers.emplace_back(hash_key, indexed_state.first);
+            }
+        }
+
+        static constexpr auto cmp_by_key = [](const auto& left, const auto& right) {
+            return left.first > right.first;
+        };
+        std::sort(best_peers.begin(), best_peers.end(), cmp_by_key);
+
+        std::set<NodeId> new_fanout_candidates;
+        auto it = best_peers.begin();
+        for (size_t i = 0; i < targets_size && it != best_peers.end(); ++i, ++it) {
+            new_fanout_candidates.insert(it->second);
+        }
+
+        tx_fanout_targets_cache_data.emplace(wtxid, new_fanout_candidates);
+        // Replace the oldest cache item with this new one.
+        if (tx_fanout_targes_cache_order.size () == 3000) {
+            auto expired_tx = tx_fanout_targes_cache_order.front();
+            tx_fanout_targets_cache_data.erase(expired_tx);
+            tx_fanout_targes_cache_order.pop_front();
+            tx_fanout_targes_cache_order.push_back(wtxid);
```

Comment:
I don't think this is right(?), this logic is never triggered.

You are checking whether the size of `tx_fanout_targes_cache_order` is over the limit, but you are only adding elements to the queue within that same `if` context, meaning that the queue is always empty.